airbyte-commons/src/main/java/io/airbyte/commons/io/ArchiveHelper.java

```diff
@@ -0,0 +1,126 @@
+
+  /**
+   * Compress a @param sourceFolder into a Gzip Tarball @param archiveFile
+   */
+  static public void createArchive(final Path sourceFolder, final Path archiveFile) throws IOException {
+    final TarArchiveOutputStream archive =
+        new TarArchiveOutputStream(new GZIPOutputStream(new BufferedOutputStream(new FileOutputStream(archiveFile.toFile()))));
+    Files.walkFileTree(sourceFolder, new SimpleFileVisitor<>() {
+
+      @Override
+      public FileVisitResult visitFile(Path file, BasicFileAttributes attributes) {
+        // only copy files, no symbolic links
+        if (attributes.isSymbolicLink()) {
+          return FileVisitResult.CONTINUE;
+        }
+        Path targetFile = sourceFolder.relativize(file);
+        try {
+          compressFile(file, targetFile, archive);
+        } catch (IOException e) {
+          LOGGER.error(String.format("Failed to archive file %s: %s", file, e));
+          throw new RuntimeException(e);
+        }
+        return FileVisitResult.CONTINUE;
+      }
+
+      @Override
+      public FileVisitResult visitFileFailed(Path file, IOException exc) {
+        LOGGER.error(String.format("Failed to include file %s in archive", file));
+        return FileVisitResult.CONTINUE;
+      }
+
+    });
+    archive.close();
+  }
+
+  static private void compressFile(final Path file, final Path filename, final TarArchiveOutputStream archive) throws IOException {
+    final TarArchiveEntry tarEntry = new TarArchiveEntry(file.toFile(), filename.toString());
+    archive.putArchiveEntry(tarEntry);
+    Files.copy(file, archive);
+    archive.closeArchiveEntry();
+  }
+
+  /**
+   * Uncompress a Gzip Tarball @param archiveFile into the @param destinationFolder
+   */
+  static public void openArchive(final Path destinationFolder, final Path archiveFile) throws IOException {
+    final TarArchiveInputStream archive =
+        new TarArchiveInputStream(new GzipCompressorInputStream(new BufferedInputStream(Files.newInputStream(archiveFile))));
+    ArchiveEntry entry;
+    while ((entry = archive.getNextEntry()) != null) {
+      final Path newPath = zipSlipProtect(entry, destinationFolder);
+      if (entry.isDirectory()) {
+        Files.createDirectories(newPath);
+      } else {
+        final Path parent = newPath.getParent();
+        if (parent != null) {
+          if (Files.notExists(parent)) {
+            Files.createDirectories(parent);
+          }
+        }
+        Files.copy(archive, newPath, StandardCopyOption.REPLACE_EXISTING);
+      }
+    }
+  }
+
+  private static Path zipSlipProtect(ArchiveEntry entry, Path destinationFolder)
+      throws IOException {
+    final Path targetDirResolved = destinationFolder.resolve(entry.getName());
+    // make sure normalized file still has destinationFolder as its prefix,
+    // else throws exception
+    final Path normalizePath = targetDirResolved.normalize();
```

Comment:
why do we need to do this normalization? what are trying to fix here?

Reply:
I think this is related to the security vulnerability here: https://nakedsecurity.sophos.com/2018/06/06/the-zip-slip-vulnerability-what-you-need-to-know/

Reply:
huh. interesting. thanks.